### ISSUE TITLE
[FR] - Ping versions should return json as well

### ISSUE THREAD

**Issue by papacarp** [input-output-hk/cardano-cli#197]
*External*  

**Area**
*Other*   ping

**Describe the feature you'd like**
Cardano ping with the json flag should return proper json.
```
./cardano-cli ping -h relay0.bluecheesestakehouse.com -p 5000 -Q -j -q
47.196.34.59:5000 Queried versions [NodeToNodeVersionV11 764824073 InitiatorAndResponder,NodeToNodeVersionV10 764824073 InitiatorAndResponder,NodeToNodeVersionV9 764824073 InitiatorAndResponder,NodeToNodeVersionV8 764824073 InitiatorAndResponder,NodeToNodeVersionV7 764824073 InitiatorAndResponder]
ubuntu@v2202308204272235639:~$ ./cardano-cli --version
cardano-cli 8.5.0.0 - linux-x86_64 - ghc-8.10
git rev a29ee68ba2c850cd50be39a3105ef191cfbc41d5
```

**Describe alternatives you've considered**
I can parse this but much better if we just have a json response option.

## Roadmap to fix

* [X] Merge https://github.com/input-output-hk/ouroboros-network/pull/4682
* [x] Merge https://github.com/input-output-hk/cardano-cli/pull/411

Once the two above are ticked, the fix will be publicly available in next `cardano-cli` release, i.e. anything newer than cardano-cli-8.13.0.0

**Comment by smelc** [comment]
Hello @papacarp :wave: New cardano-cli developer here :bow: I have been looking at the `cardano-cli ping` part of your report, and this leads me to two questions:

1/ In the case of flags `-Q -j -q`, what json do you exactly expect? Could you give an example?

2/ In the case of solely, `-j`, I'm observing:

```shell
> cardano-cli ping -m 4 -u $CARDANO_NODE_SOCKET_PATH -j
/home/churlin/dev/sanchonet/node.socket network rtt: 0.000
/home/churlin/dev/sanchonet/node.socket handshake rtt: 0.000297299s
/home/churlin/dev/sanchonet/node.socket Negotiated version NodeToClientVersionV16 4
] }
```

In the second case, would you expected something along this:

```json
{ "network_rtt": 0.000,
  "handshake_rtt": 0.00123,
  "negoatiated_version": "NodeToClientVersionV16 4" }
```

?

**Comment by smelc** [comment]
Alright, first of all, this is weird:

```shell
> cardano-cli ping -m 4 -u $CARDANO_NODE_SOCKET_PATH -j -q
] }
```

According to `-q`, only JSON should be printed. I'll look at this first.

The spurious display of `] }` likely comes from this location: https://github.com/input-output-hk/ouroboros-network/blob/cdd54f98c6743f5fd7bfe0d5a9d0b0f8a4a10b73/cardano-ping/src/Cardano/Network/Ping.hs#L102

**Comment by smelc** [comment]
The spurious display of `] }` likely comes from this location: https://github.com/input-output-hk/ouroboros-network/blob/cdd54f98c6743f5fd7bfe0d5a9d0b0f8a4a10b73/cardano-ping/src/Cardano/Network/Ping.hs#L102

**Comment by smelc** [comment]
> Alright, first of all, this is weird:
>
> ```
> > cardano-cli ping -m 4 -u $CARDANO_NODE_SOCKET_PATH -j -q
> ] }
> ```

@newhoggy> this is caused by [this lonely call to `LogEnd`](https://github.com/input-output-hk/cardano-cli/blob/10cb2a749beba465d9d7db2a4a1df082e9b4ed1b/cardano-cli/src/Cardano/CLI/Run/Ping.hs#L98). At this point, we're not sure that at least one `LogMsg` was logged so far, causing the discrepancy in case no `LogMsg` was emitted.

@newhoggy> this leads me to two questions:

1. Could we limit ourselves to write valid JSON _on a given line_. I.e. instead of expecting that the entire output is valid json (I mean `{ "pongs": [pong1, pong2, ...]}`), could we tell to users that every individual line (so `pong1`, `pong2`, etc.) is valid json? In this way, we wouldn't have to emit the `LogEnd` message at the very end. And also, I suppose that consumers of repetitive pings (I mean `-c n` with `n > 1`) are going to read the output as a stream, not at once once the call is finished, so this shouldn't make a difference for them.
2. Could we separate the handling of `-Q` and the rest in [ouroboros-network's pingClient](https://github.com/input-output-hk/ouroboros-network/blob/cdd54f98c6743f5fd7bfe0d5a9d0b0f8a4a10b73/cardano-ping/src/Cardano/Network/Ping.hs#L478)? I see that they share the same start (roughly until [here](https://github.com/input-output-hk/ouroboros-network/blob/cdd54f98c6743f5fd7bfe0d5a9d0b0f8a4a10b73/cardano-ping/src/Cardano/Network/Ping.hs#L523)) but then the rest (`keepAlive`) is only concerned with ping. Note that this is a more major change than the one in 1. so can be postponed.

**Comment by smelc** [comment]
I'm stopping for now, but did push some draft test of `ping` here: https://github.com/input-output-hk/cardano-node/commits/smelc%40improve-ping-command

**Comment by papacarp** [comment]
> Hello @papacarp 👋 New cardano-cli developer here 🙇 I have been looking at the `cardano-cli ping` part of your report, and this leads me to two questions:
> 
> 1/ In the case of flags `-Q -j -q`, what json do you exactly expect? Could you give an example?
> 
> 2/ In the case of solely, `-j`, I'm observing:
> 
> ```shell
> > cardano-cli ping -m 4 -u $CARDANO_NODE_SOCKET_PATH -j
> /home/churlin/dev/sanchonet/node.socket network rtt: 0.000
> /home/churlin/dev/sanchonet/node.socket handshake rtt: 0.000297299s
> /home/churlin/dev/sanchonet/node.socket Negotiated version NodeToClientVersionV16 4
> ] }
> ```
> 
> In the second case, would you expected something along this:
> 
> ```json
> { "network_rtt": 0.000,
>   "handshake_rtt": 0.00123,
>   "negoatiated_version": "NodeToClientVersionV16 4" }
> ```
> 
> ?

Yes, in the json version it should output proper json like you show.  Not sure about the other flags as I don't use them.  From the rest of your notes on this it seems like you are on the right track.  Please do keep it in play as it would be great to get proper json from this call.

**Comment by smelc** [comment]
> Yes, in the json version it should output proper json like you show. Not sure about the other flags as I don't use them. From the rest of your notes on this it seems like you are on the right track. Please do keep it in play as it would be great to get proper json from this call.

@papacarp> thanks for your inputs. I have to stash this work right now, but will likely be able to resume working on it later on :+1:

**Comment by smelc** [comment]
Alright, I found some time to make progress: https://github.com/input-output-hk/ouroboros-network/commits/smelc/cardano-node-fix-issue-197

Here is how the output is looking on this branch:

```shell
→ cardano-cli -- ping -m 4 -u $CARDANO_NODE_SOCKET_PATH -j
/home/churlin/dev/sanchonet/node.socket {"network_rtt":0.000}
/home/churlin/dev/sanchonet/node.socket {"handshake_rtt":0.130}
/home/churlin/dev/sanchonet/node.socket {"negotiated_version":{"initiator":"InitiatorOnly","magic":764824073,"version":"NodeToNodeVersionV11"} }
```

with `-Q`:

```shell
→ cabal run cardano-cli -- ping -h relay0.bluecheesestakehouse.com -j -p 5000 -Q
47.196.34.59:5000 {"network_rtt":0.123}
47.196.34.59:5000 {"handshake_rtt":0.122}
47.196.34.59:5000 {"queried_versions":[{"initiator":"InitiatorAndResponder","magic":764824073,"version":"NodeToNodeVersionV11"},{"initiator":"InitiatorAndResponder","magic":764824073,"version":"NodeToNodeVersionV10"},{"initiator":"InitiatorAndResponder","magic":764824073,"version":"NodeToNodeVersionV9"},{"initiator":"InitiatorAndResponder","magic":764824073,"version":"NodeToNodeVersionV8"},{"initiator":"InitiatorAndResponder","magic":764824073,"version":"NodeToNodeVersionV7"}] }
```

@papacarp> how does that look? Fine for you? I've set the floats (`network_rtt`, `handshake_rtt`) representing seconds to precision up to 3 digits after the comma.

**Comment by papacarp** [comment]
looks much better.  Is that three separate responses for one ping though?  would I need to throw away the first two responses and wait for the 3rd?  Ideally it would just give me one response with all that data.

**Comment by smelc** [comment]
@papacarp> this is the start of the stream, before the list of `pongs` starts. So in case `-Q` is omitted, these outputs are then followed by the list of pongs. I didn't yet change the format of `pongs` (which I believe could be improved), because I first need to convince the maintainer of https://github.com/input-output-hk/ouroboros-network/ to merge my changes :slightly_smiling_face: 

If you are asking whether I could smash these 3 answers into one, well, yes an no:

I could but these 3 answers are not obtained at the same time at the `ouroboros-network` level. I'd need a bigger refactoring to output only 1 aggregated answer. I would start with this smaller improvement first to see if we can get it released :slightly_smiling_face: So for now I did a minimal change, as I suspect getting a minimal change merged (see [list of pending PRs](https://github.com/input-output-hk/ouroboros-network/pulls)) is going to take a while already.
